### PR TITLE
fix: date_trunc support timezone

### LINKED ISSUE #5976: `date_trunc` drops timezone
### Describe the bug

`date_trunc` drops timezone while the input is timestamptz

### To Reproduce

```bash
❯ select date_trunc('hour', timestamptz '2000-01-01T00:00:00Z');
+------------------------------------------------------+
| datetrunc(Utf8("hour"),Utf8("2000-01-01T00:00:00Z")) |
+------------------------------------------------------+
| 2000-01-01T00:00:00                                  |
+------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
```

### Expected behavior

2000-01-01T00:00:00Z

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #5976

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/built_in_function.rs [hunk 1]
```diff
@@ -895,15 +894,37 @@ impl BuiltinScalarFunction {
...
-                vec![
-                    Exact(vec![Utf8, Timestamp(Nanosecond, None)]),
-                    Exact(vec![Utf8, Timestamp(Microsecond, None)]),
-                    Exact(vec![Utf8, Timestamp(Millisecond, None)]),
-                    Exact(vec![Utf8, Timestamp(Second, None)]),
-                ],
-                self.volatility(),
-            ),
+            BuiltinScalarFunction::DateTrunc => {
+                let time_zones = vec![
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:898]: Maybe @waitingkuo  has some ideas about how to deal with timezones in function signatures. 

It seems not quite right to have to list all timezones specially for `date_trun` only -- shouldn't the timezone logic apply to any function that takes timezones ? 🤔

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -557,13 +557,12 @@ impl BuiltinScalarFunction {
...
             BuiltinScalarFunction::DatePart => Ok(Float64),
             BuiltinScalarFunction::DateBin | BuiltinScalarFunction::DateTrunc => {
-                match input_expr_types[1] {
-                    Timestamp(Nanosecond, _) | Utf8 | Null => {
-                        Ok(Timestamp(Nanosecond, None))
-                    }
-                    Timestamp(Microsecond, _) => Ok(Timestamp(Microsecond, None)),
-                    Timestamp(Millisecond, _) => Ok(Timestamp(Millisecond, None)),
-                    Timestamp(Second, _) => Ok(Timestamp(Second, None)),
+                match &input_expr_types[1] {
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:560]: I wonder if any of the calculations actually have to change based on timezone (as wouldn't the end of the`day` being truncated depend on the timezone)?

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H  

I think this PR preserves the timezone part of date_trunc which is good. However, it doens't seem to actually change date_trunc to take the timezones into account during the calculation.

File: datafusion/expr/src/built_in_function.rs (hunk 1 above)
Comment (Weijun-H) [reply datafusion/expr/src/built_in_function.rs:898]: I am not sure there is a better way to provide the timezone here. And wait for @waitingkuo suggestion.

Also, it should be extended to other functions using timezone.

Comment (waitingkuo) [reply datafusion/expr/src/built_in_function.rs:898]: Hi @alamb  @Weijun-H @tustvold . i don't have the solution. actually i have some other pr blocked by this for now. 
i wonder whether force it to be coerced to  UTC ```Timestamp(someunit, Some("+00:00".into())``` makes sense before we have better solution for what @tustvold  suggested

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 3]
```diff
@@ -1281,7 +1281,14 @@ SELECT
...
 true false true true
 
+# date_trunc with timestamptz
+statement ok
+set timezone to '+08:00';
 
+query P
+select date_trunc('hour', timestamptz '2000-01-01T00:00:00');
+----
+2000-01-01T00:00:00+08:00
```
Comment (waitingkuo) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:1291]: i think we should a test case that truncate the day

File: datafusion/expr/src/built_in_function.rs (hunk 2 above)
Comment (waitingkuo) [reply datafusion/expr/src/built_in_function.rs:560]: https://github.com/apache/arrow-datafusion/blob/02a470f6061cce8ee8e57f7af8a6a0e0ddc1571b/datafusion/physical-expr/src/datetime_expressions.rs#L343-L347

it truncates before applying timezone offset, which is inconsistent with postgresql

```bash
❯ set timezone to '+08:00';
0 rows in set. Query took 0.026 seconds.

❯ select timestamptz '2000-01-01T00:00:00';
+-----------------------------+
| Utf8("2000-01-01T00:00:00") |
+-----------------------------+
| 2000-01-01T00:00:00+08:00   |
+-----------------------------+
1 row in set. Query took 0.002 seconds.

❯ select date_trunc('day', timestamptz '2000-01-01T00:00:00');
+-----------------------------------------------------+
| date_trunc(Utf8("day"),Utf8("2000-01-01T00:00:00")) |
+-----------------------------------------------------+
| 1999-12-31T00:00:00                                 |
+-----------------------------------------------------+
1 row in set. Query took 0.032 seconds.
```

```bash
willy=# select timestamptz '2000-01-01T00:00:00';
      timestamptz       
------------------------
 2000-01-01 00:00:00+08
(1 row)

willy=# select date_trunc('day', timestamptz '2000-01-01T00:00:00');
       date_trunc       
------------------------
 2000-01-01 00:00:00+08
(1 row)
```

**Comment by alamb** [comment]
What is the current status of this PR? It seems to have gone stale

**Comment by Weijun-H** [comment]
> What is the current status of this PR? It seems to have gone stale

This pr could truncate the day and hour now. I am not sure there is a better way to mention the timezone.

> It seems not quite right to have to list all timezones specially for date_trun only -- shouldn't the timezone logic apply to any function that takes timezones ? 🤔

**Comment by alamb** [comment]
I think @wiedld has a solution in this PR:  https://github.com/apache/arrow-datafusion/pull/7614

**Pull request closed without merging**